### DISCUSSION TITLE
Is the BusinessListBase structure specifically Fetch method signature different from 7.0 to 8.0?

### DISCUSSION THREAD

**Discussion by jhuerta3** [MarimerLLC/csla#3760] [Questions] [answered]
Hello Im working on creating a BusinessListBase class but for some reason I get errors. Im using CSLA 7.0 but the sample was created using version 8.0 .

The error that I get is "TargetParameterCountException"

Im using the sample code from the ProjectTracker sample. 

    [FetchChild]
    private void Fetch(int projectId, [Inject] IAssignmentDal dal, [Inject] IChildDataPortal<ProjectResourceEdit> portal)
    {
      var data = dal.FetchForProject(projectId);
      using (LoadListMode)
      {
        foreach (var item in data)
          Add(portal.FetchChild(item));
      }
    }

**Comment by jhuerta3** [comment]
My error

Csla.DataPortalException
  HResult=0x80131500
  Message=DataPortal.Fetch failed (Csla.BusinessListBase`2[[PCDL_2023.Shared.ParcelOldEditList, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32))
  Source=Csla
  StackTrace:
   at Csla.Reflection.ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError)
   at Csla.Server.DataPortal.Fetch(Type objectType, Object criteria, DataPortalContext context, Boolean isSync)
   at Csla.DataPortal`1.Fetch(Type objectType, Object criteria)
   at Csla.DataPortal`1.Fetch(Object[] criteria)
   at PCDL_2023.Pages.Mapping.ParcelChangeHeader.<LoadOldAndNewParcelDataForEditAsync>d__75.MoveNext() in C:\MyProjects\VS2022\PCDL_2023\PCDL_2023.Web\Pages\Mapping\ParcelChangeHeader.razor:line 688

  This exception was originally thrown at this call stack:

Inner Exception 1:
TargetParameterCountException: PCDL_2023.Shared.ParcelOldEditList.[Fetch](Int32)

Inner Exception 2:
TargetParameterCountException: Csla.BusinessListBase`2[[PCDL_2023.Shared.ParcelOldEditList, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32)

Inner Exception 3:
TargetParameterCountException: Csla.Core.ObservableBindingList`1[[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32)

Inner Exception 4:
TargetParameterCountException: Csla.Core.MobileObservableCollection`1[[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32)

Inner Exception 5:
TargetParameterCountException: System.Collections.ObjectModel.ObservableCollection`1[[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32)

Inner Exception 6:
TargetParameterCountException: System.Collections.ObjectModel.Collection`1[[PCDL_2023.Shared.ParcelOldTestEdit, PCDL_2023.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].[Fetch](Int32)

Inner Exception 7:
TargetParameterCountException: System.Object.[Fetch](Int32)

**Reply by ossendorf-at-hoelscher** [reply]
Could you share the calling code where the actual `Fetch` is called?

**Comment by jhuerta3** [comment]
So I found what is causing my issue. 

If I change the attribute for the method from [FetchChild] to [Fetch] it works(its actully hits the method in my businesslistbase class).  With the attribute of FetchChild it will not hit the method at all and I get the errors above. 

This is on the my Business List Base class which has the following:

```c#
    public class ParcelOldEditList : BusinessListBase<ParcelOldEditList, ParcelOldEdit>
{
 //along with all other method. Remove Contains Assign

         [Fetch]
        private void Fetch(string parcelChangeID, [Inject] DataAccess.DTO.IParcelOldDal dal, [Inject] IChildDataPortal<ParcelOldEdit> portal)
        {
            var data = dal.GetParcelOldListAsync(parcelChangeID);
            using (LoadListMode)
            {
                foreach (var item in data.Result)
                    Add(portal.FetchChild(item));
            }
        }

}
```

**Reply by rockfordlhotka** [reply, marked as answer]
Is the list a root object? If so, then `Fetch` is the correct attribute.

If you intend for the list to be a child object, it should be loaded as part of its parent, and the parent should use an `IChildDataPortal<C>` to fetch the child list.

**Reply by jhuerta3** [reply]
Hello @rockfordlhotka,
Yes the list is of type root object. In my sample its a 
`public class ParcelOldEdit : BusinessBase<ParcelOldEdit>`

Thank you.
